src/flask_wtf/csrf.py

```diff
@@ -77,42 +77,112 @@ def validate_csrf(data, secret_key=None, time_limit=None, token_key=None):
 
     :raises ValidationError: Contains the reason that validation failed.
 
+    .. versionchanged:: 1.21.post1
+        Fallbacks to legacy_validate_csrf method. This provides a
+        comptability layer for old clients.
     .. versionchanged:: 0.14
         Raises ``ValidationError`` with a specific error message rather than
         returning ``True`` or ``False``.
     """
+    try:
+        secret_key = _get_config(
+            secret_key,
+            "WTF_CSRF_SECRET_KEY",
+            current_app.secret_key,
+            message="A secret key is required to use CSRF.",
+        )
+        field_name = _get_config(
+            token_key,
+            "WTF_CSRF_FIELD_NAME",
+            "csrf_token",
+            message="A field name is required to use CSRF.",
+        )
+        time_limit = _get_config(time_limit, "WTF_CSRF_TIME_LIMIT", 3600, required=False)
 
-    secret_key = _get_config(
-        secret_key,
-        "WTF_CSRF_SECRET_KEY",
-        current_app.secret_key,
-        message="A secret key is required to use CSRF.",
-    )
-    field_name = _get_config(
-        token_key,
-        "WTF_CSRF_FIELD_NAME",
-        "csrf_token",
-        message="A field name is required to use CSRF.",
-    )
-    time_limit = _get_config(time_limit, "WTF_CSRF_TIME_LIMIT", 3600, required=False)
+        if not data:
+            raise ValidationError("The CSRF token is missing.")
+
+        if field_name not in session:
+            raise ValidationError("The CSRF session token is missing.")
 
-    if not data:
-        raise ValidationError("The CSRF token is missing.")
+        s = URLSafeTimedSerializer(secret_key, salt="wtf-csrf-token")
 
-    if field_name not in session:
-        raise ValidationError("The CSRF session token is missing.")
+        try:
+            token = s.loads(data, max_age=time_limit)
+        except SignatureExpired as e:
+            raise ValidationError("The CSRF token has expired.") from e
+        except BadData as e:
+            raise ValidationError("The CSRF token is invalid.") from e
+
+        if not hmac.compare_digest(session[field_name], token):
+            raise ValidationError("The CSRF tokens do not match.")
+    except Exception as e:
+        logger.info("Falling back to legacy CSRF validation.")
+        token_key = 'csrf_token' if token_key is None else token_key
+        is_valid = legacy_validate_csrf(
+            data=data,
+            secret_key=secret_key,
+            time_limit=time_limit,
+            token_key=token_key
+        )
+        if is_valid is False:
+            raise e
```

Comment:
kinda not sure this is the right error in all cases... if it doesn't validate with the new version, but it's actually an expired/invalid old kind of token, we should actually fail with "old-style validation error" i think? the only time we actually want to attempt "legacy validation" is if `URLSafeTimedSerializer(secret_key, salt="wtf-csrf-token").loads(data)` fails? trying to solidify my understanding of error cases + ensure we have clear branch flow; the `except` at the end makes it hard to tell why we might reach there, even though i'm pretty sure there will be no false positives or negatives with this implementation

Reply:
so the old "legacy validation" has a different API than new validation -- which makes it hard to return the "legacy validation" error.

- The new validation will raise a ValidationError if invalid. Otherwise nothing.
- The old validation will return true or false.

So if we returned the old result (`true` or `false`) -- then those would both be seen as a valid result from anything that's consuming the output of the new validation.

From the caller's perspective, I think they can just treat this forked function as the same as the unforked new validation. It will either raise a known `ValidationError` if invalid or it will not.


writing this out, I guess this would be better accomplished by wrapping `legacy_validate_csrf()` in another `try..catch`, so we don't leak any of those errors as well.